tests/instruments/lecroy/test_lecroyT3DSO1204.py

```diff
@@ -0,0 +1,469 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import pytest
+
+from pymeasure.instruments.lecroy.lecroyT3DSO1204 import LeCroyT3DSO1204
+from pymeasure.test import expected_protocol
+
+INVALID_CHANNELS = ["INVALID_SOURCE", "C1 C2", "C1 MATH", "C1234567", "LINE", "CHANNEL"]
+VALID_CHANNELS = [('C1', 1), ('CHANNEL2', 2), ('ch 3', 3), ('chan 4', 4), ('\tC3\t', 3),
+                  (' math ', 'MATH')]
+
+
+def test_init():
+    with expected_protocol(
+            LeCroyT3DSO1204,
+            [(b"CHDR OFF", None)]
+    ):
+        pass  # Verify the expected communication.
+
+
+@pytest.mark.parametrize("channel", INVALID_CHANNELS)
+def test_invalid_source(channel):
+    with pytest.raises(ValueError):
+        with expected_protocol(LeCroyT3DSO1204, [(b"CHDR OFF", None)]) as instr:
+            instr.ch(channel)
+
+
+@pytest.mark.parametrize("channel", VALID_CHANNELS)
+def test_sanitize_valid_source(channel):
+    with expected_protocol(LeCroyT3DSO1204, [(b"CHDR OFF", None)]) as instr:
+        if isinstance(channel[1], int):
+            assert channel[1] == instr.ch(channel[0]).number
+        else:
+            assert instr == instr.ch(channel[0])
```

Comment:
A remark what I mentioned earlier:

It is good, that you test the `instr.ch` method.
However, I find it useful to already test the functions for themselves, in this case test `sanitize` itself.
That way you do reduce the complexity (no working with instruments and several functions) and you know, that this single piece of code (the `sanitize` function) works as expected.

Just as a suggestion, also for the future. External functions are easy to test and the tests give confidence, that this external piece of code works as expected.

Reply:
That makes sense. I applied your suggestions in commit c96323c1187e87d8cc82b658f1a7cfaf5a62848e